### PR TITLE
Add generic Controller for Canopen

### PR DESCRIPTION
**Pull request by destogl**
The controller enables sending of generic command to can open devices.
It is not intended for the end-use, but for prototyping and debugging.

- [x] Check documentation on how to use and add missing parts (done in #56)

### CODE REVIEW

File: canopen_ros2_controllers/src/canopen_proxy_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+static constexpr rmw_qos_profile_t rmw_qos_profile_services_hist_keep_all = {
+  RMW_QOS_POLICY_HISTORY_KEEP_ALL,
+  1,  // message queue depth
+  RMW_QOS_POLICY_RELIABILITY_RELIABLE,
+  RMW_QOS_POLICY_DURABILITY_VOLATILE,
+  RMW_QOS_DEADLINE_DEFAULT,
+  RMW_QOS_LIFESPAN_DEFAULT,
+  RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
+  RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
+  false};
```
Comment (ipa-cmh) [inline canopen_ros2_controllers/src/canopen_proxy_controller.cpp:39]: Probably remove this.

```suggestion
```

File: canopen_ros2_controllers/test/test_load_canopen_proxy_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2022, Stogl Robotics Consulting UG (haftungsbeschränkt)
+// Copyright (c) 2022, Stogl Robotics Consulting UG (haftungsbeschränkt) (template)
```
Comment (destogl) [inline canopen_ros2_controllers/test/test_load_canopen_proxy_controller.cpp:2]:
```suggestion
```

File: canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+using canopen_ros2_controllers::CMD_MY_ITFS;
+using canopen_ros2_controllers::STATE_MY_ITFS;
+
+class CanopenProxyControllerTest : public CanopenProxyControllerFixture<TestableCanopenProxyController>
```
[the remark this answers is not in the document]
Comment (livanov93) [reply canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp:26]: They are commented out, so not in use.

File: canopen_ros2_controllers/src/canopen_proxy_controller.cpp [hunk 4]
```diff
@@ -25,18 +24,6 @@
...
-static constexpr rmw_qos_profile_t rmw_qos_profile_services_hist_keep_all = {
-  RMW_QOS_POLICY_HISTORY_KEEP_ALL,
-  1,  // message queue depth
-  RMW_QOS_POLICY_RELIABILITY_RELIABLE,
-  RMW_QOS_POLICY_DURABILITY_VOLATILE,
-  RMW_QOS_DEADLINE_DEFAULT,
-  RMW_QOS_LIFESPAN_DEFAULT,
-  RMW_QOS_POLICY_LIVELINESS_SYSTEM_DEFAULT,
-  RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT,
-  false};
```
Comment (livanov93) [inline canopen_ros2_controllers/src/canopen_proxy_controller.cpp]: @destogl We keep it or shoud I remove it?

**Comment by livanov93** [comment]
@destogl @ipa-cmh 
Please test [4d7a6a7](https://github.com/ros-industrial/ros2_canopen/pull/54/commits/4d7a6a7ac3581ccc793f098c3cb23d967aca79f8) with:
```
colcon test --packages-select canopen_ros2_controllers --event-handlers desktop_notification+ status+ summary- console_direct+ terminal_title+ --ctest-args -R test_canopen_proxy_controller 
```
and

```
colcon test --packages-select canopen_ros2_controllers --event-handlers desktop_notification+ status+ summary- console_direct+ terminal_title+ --ctest-args -R test_load_canopen_proxy_controller 
```

**Comment by ipa-cmh** [comment]
For the first one I get the following result - 1 test failed:

```
colcon test --packages-select canopen_ros2_controllers --event-handlers desktop_notification+ status+ summary- console_direct+ terminal_title+ --ctest-args -R test_canopen_proxy_controller 
Starting >>> canopen_ros2_controllers
UpdateCTestConfiguration  from :/home/christoph/ws_ros2/build/canopen_ros2_controllers/CTestConfiguration.ini
Parse Config file:/home/christoph/ws_ros2/build/canopen_ros2_controllers/CTestConfiguration.ini
   Site: christoph-ThinkPad-T490
   Build name: (empty)
 Add coverage exclude regular expressions.
Create new tag: 20220920-1357 - Experimental
UpdateCTestConfiguration  from :/home/christoph/ws_ros2/build/canopen_ros2_controllers/CTestConfiguration.ini
Parse Config file:/home/christoph/ws_ros2/build/canopen_ros2_controllers/CTestConfiguration.ini
Test project /home/christoph/ws_ros2/build/canopen_ros2_controllers
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 2
    Start 2: test_canopen_proxy_controller

2: Test command: /usr/bin/python3.10 "-u" "/opt/ros/humble/share/ament_cmake_test/cmake/run_test.py" "/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_canopen_proxy_controller.gtest.xml" "--package-name" "canopen_ros2_controllers" "--output-file" "/home/christoph/ws_ros2/build/canopen_ros2_controllers/ament_cmake_gmock/test_canopen_proxy_controller.txt" "--command" "/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_canopen_proxy_controller" "--gtest_output=xml:/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_canopen_proxy_controller.gtest.xml"
2: Test timeout computed to be: 60
2: -- run_test.py: invoking following command in '/home/christoph/ws_ros2/build/canopen_ros2_controllers':
2:  - /home/christoph/ws_ros2/build/canopen_ros2_controllers/test_canopen_proxy_controller --gtest_output=xml:/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_canopen_proxy_controller.gtest.xml
2: Running main() from gmock_main.cc
2: [==========] Running 10 tests from 2 test suites.
2: [----------] Global test environment set-up.
2: [----------] 9 tests from CanopenProxyControllerTest
2: [ RUN      ] CanopenProxyControllerTest.joint_names_parameter_not_set
2: [ERROR] [1663682254.699946999] [test_canopen_ros2_controllers]: 'joint' parameter was empty
2: [       OK ] CanopenProxyControllerTest.joint_names_parameter_not_set (15 ms)
2: [ RUN      ] CanopenProxyControllerTest.all_parameters_set_configure_success
2: [INFO] [1663682254.711912406] [test_canopen_ros2_controllers]: configure successful
2: [       OK ] CanopenProxyControllerTest.all_parameters_set_configure_success (14 ms)
2: [ RUN      ] CanopenProxyControllerTest.check_exported_intefaces
2: [INFO] [1663682254.724553535] [test_canopen_ros2_controllers]: configure successful
2: [       OK ] CanopenProxyControllerTest.check_exported_intefaces (12 ms)
2: [ RUN      ] CanopenProxyControllerTest.activate_success
2: [INFO] [1663682254.737476876] [test_canopen_ros2_controllers]: configure successful
2: [       OK ] CanopenProxyControllerTest.activate_success (14 ms)
2: [ RUN      ] CanopenProxyControllerTest.update_success
2: [INFO] [1663682254.751703158] [test_canopen_ros2_controllers]: configure successful
2: [       OK ] CanopenProxyControllerTest.update_success (13 ms)
2: [ RUN      ] CanopenProxyControllerTest.deactivate_success
2: [INFO] [1663682254.764671982] [test_canopen_ros2_controllers]: configure successful
2: [       OK ] CanopenProxyControllerTest.deactivate_success (13 ms)
2: [ RUN      ] CanopenProxyControllerTest.reactivate_success
2: [INFO] [1663682254.777885047] [test_canopen_ros2_controllers]: configure successful
2: [       OK ] CanopenProxyControllerTest.reactivate_success (14 ms)
2: [ RUN      ] CanopenProxyControllerTest.test_setting_slow_mode_service
2: [INFO] [1663682254.792046905] [test_canopen_ros2_controllers]: configure successful
2: [       OK ] CanopenProxyControllerTest.test_setting_slow_mode_service (13 ms)
2: [ RUN      ] CanopenProxyControllerTest.test_update_logic_fast
2: [INFO] [1663682254.804379992] [test_canopen_ros2_controllers]: configure successful
2: [       OK ] CanopenProxyControllerTest.test_update_logic_fast (12 ms)
2: [----------] 9 tests from CanopenProxyControllerTest (121 ms total)
2: 
2: [----------] 1 test from MissingMandatoryParameterDuringConfiguration/CanopenProxyControllerTestParameterizedParameters
2: [ RUN      ] MissingMandatoryParameterDuringConfiguration/CanopenProxyControllerTestParameterizedParameters.one_parameter_is_missing/0
2: [ERROR] [1663682254.823898670] [test_canopen_ros2_controllers]: 'joint' parameter was empty
2: [       OK ] MissingMandatoryParameterDuringConfiguration/CanopenProxyControllerTestParameterizedParameters.one_parameter_is_missing/0 (11 ms)
2: [----------] 1 test from MissingMandatoryParameterDuringConfiguration/CanopenProxyControllerTestParameterizedParameters (11 ms total)
2: 
2: [----------] Global test environment tear-down
2: [==========] 10 tests from 2 test suites ran. (145 ms total)
2: [  PASSED  ] 10 tests.
2: -- run_test.py: return code 0                       
2: -- run_test.py: inject classname prefix into gtest result file '/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_canopen_proxy_controller.gtest.xml'
2: -- run_test.py: verify result file '/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_canopen_proxy_controller.gtest.xml'
1/1 Test #2: test_canopen_proxy_controller ....   Passed    0.33 sec

The following tests passed:
	test_canopen_proxy_controller

100% tests passed, 0 tests failed out of 1

Label Time Summary:
gmock    =   0.33 sec*proc (1 test)

Total Test time (real) =   0.34 sec
Finished <<< canopen_ros2_controllers [0.40s]          
```


Second on:
```
colcon test --packages-select canopen_ros2_controllers --event-handlers desktop_notification+ status+ summary- console_direct+ terminal_title+ --ctest-args -R test_load_canopen_proxy_controller 
Starting >>> canopen_ros2_controllers
UpdateCTestConfiguration  from :/home/christoph/ws_ros2/build/canopen_ros2_controllers/CTestConfiguration.ini
Parse Config file:/home/christoph/ws_ros2/build/canopen_ros2_controllers/CTestConfiguration.ini
   Site: christoph-ThinkPad-T490
   Build name: (empty)
 Add coverage exclude regular expressions.
Create new tag: 20220920-1357 - Experimental
UpdateCTestConfiguration  from :/home/christoph/ws_ros2/build/canopen_ros2_controllers/CTestConfiguration.ini
Parse Config file:/home/christoph/ws_ros2/build/canopen_ros2_controllers/CTestConfiguration.ini
Test project /home/christoph/ws_ros2/build/canopen_ros2_controllers
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: test_load_canopen_proxy_controller

1: Test command: /usr/bin/python3.10 "-u" "/opt/ros/humble/share/ament_cmake_test/cmake/run_test.py" "/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_load_canopen_proxy_controller.gtest.xml" "--package-name" "canopen_ros2_controllers" "--output-file" "/home/christoph/ws_ros2/build/canopen_ros2_controllers/ament_cmake_gmock/test_load_canopen_proxy_controller.txt" "--command" "/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_load_canopen_proxy_controller" "--gtest_output=xml:/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_load_canopen_proxy_controller.gtest.xml"
1: Test timeout computed to be: 60
1: -- run_test.py: invoking following command in '/home/christoph/ws_ros2/build/canopen_ros2_controllers':
1:  - /home/christoph/ws_ros2/build/canopen_ros2_controllers/test_load_canopen_proxy_controller --gtest_output=xml:/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_load_canopen_proxy_controller.gtest.xml
1: Running main() from gmock_main.cc
1: [==========] Running 1 test from 1 test suite.
1: [----------] Global test environment set-up.
1: [----------] 1 test from TestLoadCanopenProxyController
1: [ RUN      ] TestLoadCanopenProxyController.load_controller
1: [INFO] [1663682236.161395986] [resource_manager]: Loading hardware 'TestActuatorHardware' 
1: [INFO] [1663682236.162179674] [resource_manager]: Initialize hardware 'TestActuatorHardware' 
1: [INFO] [1663682236.162194392] [resource_manager]: Successful initialization of hardware 'TestActuatorHardware'
1: [INFO] [1663682236.162203586] [resource_manager]: Loading hardware 'TestSensorHardware' 
1: [INFO] [1663682236.162315872] [resource_manager]: Initialize hardware 'TestSensorHardware' 
1: [INFO] [1663682236.162323729] [resource_manager]: Successful initialization of hardware 'TestSensorHardware'
1: [INFO] [1663682236.162327828] [resource_manager]: Loading hardware 'TestSystemHardware' 
1: [INFO] [1663682236.162466599] [resource_manager]: Initialize hardware 'TestSystemHardware' 
1: [INFO] [1663682236.162482388] [resource_manager]: Successful initialization of hardware 'TestSystemHardware'
1: [INFO] [1663682236.175082874] [test_controller_manager]: Loading controller 'test_canopen_ros2_controllers'
1: [       OK ] TestLoadCanopenProxyController.load_controller (28 ms)
1: [----------] 1 test from TestLoadCanopenProxyController (28 ms total)
1: 
1: [----------] Global test environment tear-down
1: [==========] 1 test from 1 test suite ran. (28 ms total)
1: [  PASSED  ] 1 test.
1: -- run_test.py: return code 0                       
1: -- run_test.py: inject classname prefix into gtest result file '/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_load_canopen_proxy_controller.gtest.xml'
1: -- run_test.py: verify result file '/home/christoph/ws_ros2/build/canopen_ros2_controllers/test_results/canopen_ros2_controllers/test_load_canopen_proxy_controller.gtest.xml'
1/1 Test #1: test_load_canopen_proxy_controller ...   Passed    0.22 sec

The following tests passed:
	test_load_canopen_proxy_controller

100% tests passed, 0 tests failed out of 1

Label Time Summary:
gmock    =   0.22 sec*proc (1 test)

Total Test time (real) =   0.22 sec
Finished <<< canopen_ros2_controllers [0.29s]

```

**Comment by livanov93** [comment]
> 100% tests passed, 0 tests failed out of 1

@ipa-cmh  I can see for both of them `100% tests passed, 0 tests failed out of 1`. Am I right?

**Comment by ipa-cmh** [comment]
@livanov93 sorry, only saw error and jumped to conclusions :D

File: canopen_ros2_controllers/src/canopen_proxy_controller.cpp (hunk 4 above)
Comment (destogl) [reply canopen_ros2_controllers/src/canopen_proxy_controller.cpp]: We can remove it. You have introduced local variable, which is also a good solution.

File: canopen_ros2_controllers/test/test_canopen_proxy_controller.hpp [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+//    }
+//    auto result = slow_control_service_client_->async_send_request(request);
+//    EXPECT_EQ(executor.spin_until_future_complete(result), rclcpp::FutureReturnCode::SUCCESS);
+//
+//    return result.get();
+//  }
+
+protected:
+  // TODO(anyone): adjust the members as needed
+
```
Comment (destogl) [inline canopen_ros2_controllers/test/test_canopen_proxy_controller.hpp:244]:
```suggestion
```

File: canopen_ros2_controllers/test/test_canopen_proxy_controller.hpp [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+  // Controller-related parameters
+  std::string joint_name_ = {"joint1"};
+//  std::vector<std::string> state_joint_names_ = {"joint1state"};
```
Comment (destogl) [inline canopen_ros2_controllers/test/test_canopen_proxy_controller.hpp:247]:
```suggestion
```

File: canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+{
+  SetUpController();
+
+  ASSERT_TRUE(controller_->joint_name_.empty());
+
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+
+  // TODO(livanov93): check if it is string
+  // hardcoded as it should be "joint1"
+  ASSERT_THAT(controller_->joint_name_, testing::HasSubstr("joint"));
```
Comment (destogl) [inline canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp:63]: Wouldn't this be better?
```suggestion
  ASSERT_THAT(controller_->joint_name_, joint_name_);
```

File: canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+  ASSERT_TRUE(std::isnan(controller_->command_interfaces_[CommandInterfaces::TPDO_DATA].get_value()));
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_TRUE(std::isnan(controller_->command_interfaces_[CommandInterfaces::TPDO_DATA].get_value()));
+
+  ASSERT_EQ(
+    controller_->update(rclcpp::Time(0), rclcpp::Duration::from_seconds(0.01)),
+    controller_interface::return_type::OK);
+}
+
+TEST_F(CanopenProxyControllerTest, test_setting_slow_mode_service)
```
Comment (destogl) [inline canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp:139]: Please rename the tests what it is actually testing

File: canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    controller_interface::return_type::OK);
+
+  EXPECT_EQ(joint_command_values_[CMD_TPDO_DATA], 0.45);
+
+  subscribe_and_get_messages(msg);
+
+  ASSERT_EQ(msg.set_point, 0.45);
+}
+
+ */
```
Comment (destogl) [inline canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp:280]: Can we remove those?
```suggestion
```

File: canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+  ASSERT_EQ(controller_->on_deactivate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+}
+
+TEST_F(CanopenProxyControllerTest, reactivate_success)
+{
+  SetUpController();
+
+  ASSERT_EQ(controller_->on_configure(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_EQ(controller_->on_activate(rclcpp_lifecycle::State()), NODE_SUCCESS);
+  ASSERT_EQ(controller_->command_interfaces_[CommandInterfaces::TPDO_DATA].get_value(), 101.101);
```
Comment (destogl) [inline canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp:128]: Can we all all interface to test their values? Maybe would also be a good idea to slightly change their value.

**Review by destogl — COMMENTED** [review]
@livanov93 Thanks for the test, just few minor comments and we can merge this from my side.

File: canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp (hunk 9 above)
Comment (livanov93) [reply canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp:280]: sure

File: canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp (hunk 10 above)
Comment (livanov93) [reply canopen_ros2_controllers/test/test_canopen_proxy_controller.cpp:128]: @destogl I planned to add more attention to the value propagation and canopen logic on 402 profile implementation.

**Comment by livanov93** [comment]
@destogl ready for merge

**Review by ipa-cmh — APPROVED** [review]

**Comment by ipa-cmh** [comment]
@destogl is merging okay, then I'll proceed.

**Review by destogl — COMMENTED** [review]
Thanks @livanov93  for following up!

@ipa-cmh you can merge this! Thanks for the patience!

**Pull request merged**